src/Common/src/Microsoft/Internal/Assumes.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Globalization;
+using System.Runtime.Serialization;
+
+namespace Microsoft.Internal
+{
+    internal static partial class Assumes
+    {
+        [DebuggerStepThrough]
+        internal static void NotNull<T>(T value)
+            where T : class
+        {
+            IsTrue(value != null);
+        }
+
+        [DebuggerStepThrough]
+        internal static void NotNull<T1, T2>(T1 value1, T2 value2)
+            where T1 : class
+            where T2 : class
+        {
+            NotNull(value1);
+            NotNull(value2);
+        }
+
+        [DebuggerStepThrough]
+        internal static void NotNull<T1, T2, T3>(T1 value1, T2 value2, T3 value3)
+            where T1 : class
+            where T2 : class
+            where T3 : class
+        {
+            NotNull(value1);
+            NotNull(value2);
+            NotNull(value3);
+        }
+
+        [DebuggerStepThrough]
+        internal static void NotNullOrEmpty(string value)
+        {
+            NotNull(value);
+            IsTrue(value.Length > 0);
+        }
+
+        [DebuggerStepThrough]
+        internal static void IsTrue(bool condition)
+        {
+            if (!condition)
+            {
+                throw UncatchableException(null);
+            }
+        }
+
+        [DebuggerStepThrough]
+        internal static void IsTrue(bool condition, /*[Localizable(false)]*/string message)
```

Comment:
> What's the plan for these commented-out ```[Localizable(...)]``` attributes?

I'm not sure- perhaps at the time this was being written the attribute wasn't portable.  Can you shine any light on this, @nblumhardt?